utils/CircuitCheck/UnitaryBuilder.cpp

```diff
@@ -211,9 +228,9 @@ void UnitaryBuilder::applyMatrix(ArrayRef<Complex> u,
 }
 
 void UnitaryBuilder::applyMatrix(ArrayRef<Complex> u, unsigned numTargets,
-                                 const std::vector<Qubit> &qubits) {
-  auto qubitsSorted = qubits;
-  std::sort(qubitsSorted.begin(), qubitsSorted.end());
+                                 ArrayRef<Qubit> qubits) {
+  SmallVector<Qubit, 16> qubitsSorted(qubits);
+  llvm::sort(qubitsSorted);
```

Comment:
Curious: what does it mean to "sort" qubits?

Reply:
Short answer: In the unitary builder, qubits are represented as integers that, not only, uniquely identify them but also locate their position in the quantum state.  The position is important to appropriately create the unitary, and the decision of whether the 0th qubit is the rightmost or the leftmost one will change the resulting unitary.

More details:
CircuitCheck consider the rightmost qubit as the 0th one. For exmple two-qubit system in the |00> state:

```
qubit:   1 0
state: | 0 0 >
```

The unitary that modifies the state of this system is mapping between states. When a build a unitary, we start with the identity:

```
        |00> |01> |10> |11>
|00>  |  1    0    0    0  |
|01>  |  0    1    0    0  |
|10>  |  0    0    1    0  |
|11>  |  0    0    0    1  |
```

Now, imagine the circuit applies a one-qubit operation to the 0th qubit, say `X`, which has the following matrix representation:

```
       |0>  |1> 
|0>  |  0    1  |
|1>  |  1    0  |
```

We "can't" directly compute the resulting unitary because the matrices have different dimensions. First, we combined this `X` matrix with an identity matrix of appropriate size, so that the resulting one will have the same dimensions as the unitary we are building, then we could multiply them. CircuitCheck, however, will not build this combined matrix explicitly, instead it will do some index dark magic to figure it out how to modify the unitary its building directly. Part of this computation involves sorting the qubits w.r.t their position in the quantum state.